### ISSUE TITLE
Database error when enabling app: 1067 Invalid default value for 'expires'

### ISSUE THREAD

**Issue by Miesepies** [ChristophWurst/twofactor_admin#24]
Nextcloud 14.0.4 docker image (official)
MariaDB docker image (official)

An exception occurred while executing 'ALTER TABLE oc_twofactor_admin_codes CHANGE `expires` expires BIGINT UNSIGNED DEFAULT NULL NOT NULL': SQLSTATE[42000]: Syntax error or access violation: 1067 Invalid default value for 'expires'

Could it be that "DEFAULT NULL NOT NULL" is the issue here?

```
MariaDB [nextcloud]> describe oc_twofactor_admin_codes;                                                                                                                           
+---------+---------------------+------+-----+---------+-------+                                                                                                                  
| Field   | Type                | Null | Key | Default | Extra |                                                                                                                  
+---------+---------------------+------+-----+---------+-------+                                                                                                                  
| id      | bigint(20) unsigned | NO   | PRI | NULL    |       |                                                                                                                  
| user_id | varchar(64)         | NO   | PRI | NULL    |       |                                                                                                                  
| code    | varchar(6)          | NO   |     | NULL    |       |                                                                                                                  
| expires | bigint(20) unsigned | YES  |     | NULL    |       |                                                                                                                  
+---------+---------------------+------+-----+---------+-------+                                                                                                                  
4 rows in set (0.002 sec)                                          

```

_Originally posted by @Miesepies in https://github.com/ChristophWurst/twofactor_admin/issues/20#issuecomment-442793190_

**Comment by ChristophWurst** [comment]
Please post the output of ``php occ migrations:status twofactor_admin``

**Comment by Ryonez** [comment]
I have the same issue while trying to install. The command you want run outputs:

![image](https://user-images.githubusercontent.com/16157130/49268094-68918480-f4c2-11e8-93b7-a6ec8546234b.png)

**Comment by ChristophWurst** [comment]
This is totally unrelated error. Please file a new ticket.

**Comment by Ryonez** [comment]
Then this is one being produced by the same error above.
I only ran the command because I have op's issue.

**Comment by ChristophWurst** [comment]
Got it. What version do you run? v0.2.0? Does https://github.com/ChristophWurst/twofactor_admin/blob/v0.2.0/lib/Migration/Version2Date20180926055748.php on your instance?

**Comment by Ryonez** [comment]
This happens on install:
![image](https://user-images.githubusercontent.com/16157130/49282338-ce4a3480-f4f3-11e8-9d09-128178eed956.png)

> Does https://github.com/ChristophWurst/twofactor_admin/blob/v0.2.0/lib/Migration/Version2Date20180926055748.php on your instance?

Are you asking if that file is present? (It is)

**Comment by Ryonez** [comment]
Opps, just noticed that it seems to be trying to install 0.1.0.

![image](https://user-images.githubusercontent.com/16157130/49283298-46b1f500-f4f6-11e8-9ec0-67e42743e25a.png)

**Comment by Miesepies** [comment]
Yes, 0.1.0 for nc14, 0.2.0 vor nc15 as seen on the app page.

![image](https://user-images.githubusercontent.com/45454908/49285251-40406080-f497-11e8-9328-be8664c22299.png)

**Comment by Ryonez** [comment]
Good to know, but isn't 15 a rc? It's not the latest stable branch.

**Comment by k0d3g3ar** [comment]
Same issue here.  NC 15 and manually installed 0.2.0.   When attempting to enable, I get this:
An exception occurred while executing 'ALTER TABLE oc_twofactor_admin_codes CHANGE `expires` expires BIGINT UNSIGNED DEFAULT NULL NOT NULL': SQLSTATE[42000]: Syntax error or access violation: 1067 Invalid default value for 'expires'

**Comment by ChristophWurst** [comment]
While I unfortunately can not yet provide a solution for this database incompatibility for Nextcloud 15 yet, I have prepared releases of twofactor_totp and twofactor_u2f that now support being disabled via the CLI as well.

Please see https://github.com/nextcloud/twofactor_u2f/pull/311 and https://github.com/nextcloud/twofactor_totp/pull/360 and let me know if that helps you unlock your or your users accounts. Thank you :v:

**Comment by MagnumGR** [comment]
> This happens on install:
> ![image](https://user-images.githubusercontent.com/16157130/49282338-ce4a3480-f4f3-11e8-9d09-128178eed956.png)
> 
> > Does https://github.com/ChristophWurst/twofactor_admin/blob/v0.2.0/lib/Migration/Version2Date20180926055748.php on your instance?
> 
> Are you asking if that file is present? (It is)

I have the same problem with NC15 on Fedora 29 when i try to install two-factor admin support.

**Comment by ricain** [comment]
Same issue with NC 15 and twofactor_admin 0.2.0.

![img_1061](https://user-images.githubusercontent.com/4070933/53081644-8be53200-34fb-11e9-9b13-bc2158d4e0dc.PNG)

**Comment by rainbob5** [comment]
Hi, i had the same issue but by changing the table structure I was able to enable the app and it works perfectly !

Here's the code:

`DROP TABLE IF EXISTS twofactor_admin_codes;
CREATE TABLE twofactor_admin_codes (
  id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  user_id varchar(64) COLLATE utf8mb4_bin NOT NULL,
  code varchar(6) COLLATE utf8mb4_bin NOT NULL,
  expires bigint(20) unsigned NOT NULL,
  PRIMARY KEY (id,user_id)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED;`

**Comment by MagnumGR** [comment]
^^
Does not work for me ...
I get the same error

**Comment by rainbob5** [comment]
> 
> 
> ^^
> Does not work for me ...
> I get the same error

What's your table structure ? You can get it with this: `DESCRIBE twofactor_admin_codes;`

**Comment by MagnumGR** [comment]
DESCRIBE oc_twofactor_admin_codes;
+---------+---------------------+------+-----+---------+----------------+
| Field   | Type                | Null | Key | Default | Extra          |
+---------+---------------------+------+-----+---------+----------------+
| id      | bigint(20) unsigned | NO   | PRI | NULL    | auto_increment |
| user_id | varchar(64)         | NO   | PRI | NULL    |                |
| code    | varchar(6)          | NO   |     | NULL    |                |
| expires | bigint(20) unsigned | NO   |     | NULL    |                |
+---------+---------------------+------+-----+---------+----------------+
4 rows in set (0.001 sec)

**Comment by MagnumGR** [comment]
IT WORKS !
After upgrading to latest NC 15 (v15.0.5) the App (Two-Factor Admin Support ) installs without error.

**Comment by telecomguyML** [comment]
> Hi, i had the same issue but by changing the table structure I was able to enable the app and it works perfectly !
> 
> Here's the code:
> 
> `DROP TABLE IF EXISTS twofactor_admin_codes; CREATE TABLE twofactor_admin_codes ( id bigint(20) unsigned NOT NULL AUTO_INCREMENT, user_id varchar(64) COLLATE utf8mb4_bin NOT NULL, code varchar(6) COLLATE utf8mb4_bin NOT NULL, expires bigint(20) unsigned NOT NULL, PRIMARY KEY (id,user_id) ) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED;`

This worked for me, although I had to change the table from twofactor_admin_codes to oc_twofactor_admin_codes.  Once I did that, I attempted to enable the app and it did so without issues.  I am running 15.0.7.

**Comment by christianlupus** [comment]
This might be solved by my PR #31.

**Comment by ChristophWurst** [comment]
Yep, this should be fixed now.
